sw/airborne/firmwares/rotorcraft/stabilization/stabilization_rate_indi.c

```diff
@@ -23,29 +23,106 @@
+#endif
+
+#define ROLL_RATE_DEADBAND_EXCEEDED()                                         \
+  (radio_control.values[RADIO_ROLL] >  STABILIZATION_RATE_DEADBAND_P || \
+   radio_control.values[RADIO_ROLL] < -STABILIZATION_RATE_DEADBAND_P)
+
+#define PITCH_RATE_DEADBAND_EXCEEDED()                                         \
+  (radio_control.values[RADIO_PITCH] >  STABILIZATION_RATE_DEADBAND_Q || \
+   radio_control.values[RADIO_PITCH] < -STABILIZATION_RATE_DEADBAND_Q)
+
+#define YAW_RATE_DEADBAND_EXCEEDED()                                         \
+  (radio_control.values[RADIO_YAW] >  STABILIZATION_RATE_DEADBAND_R || \
+   radio_control.values[RADIO_YAW] < -STABILIZATION_RATE_DEADBAND_R)
+
+#if PERIODIC_TELEMETRY
+#include "subsystems/datalink/telemetry.h"
+
+static void send_rate(struct transport_tx *trans, struct link_device *dev)
+{
+  pprz_msg_send_RATE_LOOP(trans, dev, AC_ID,
+                          &stabilization_rate_sp.p,
+                          &stabilization_rate_sp.q,
+                          &stabilization_rate_sp.r,
+                          &stabilization_cmd[COMMAND_THRUST]);
+}
+#endif
+
 void stabilization_rate_init(void)
 {
   stabilization_indi_init();
-}
 
+#if PERIODIC_TELEMETRY
+  register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_RATE_LOOP, send_rate);
+#endif
+}
 
 void stabilization_rate_read_rc(void)
 {
-  //FIXME: make a new indi function
+  if (ROLL_RATE_DEADBAND_EXCEEDED()) {
+    stabilization_rate_sp.p = (float) radio_control.values[RADIO_ROLL] * STABILIZATION_INDI_MAX_RATE / MAX_PPRZ;
+  } else {
+    stabilization_rate_sp.p = 0;
+  }
+
+  if (PITCH_RATE_DEADBAND_EXCEEDED()) {
+    stabilization_rate_sp.q = (float) radio_control.values[RADIO_PITCH] * STABILIZATION_INDI_MAX_RATE / MAX_PPRZ;
+  } else {
+    stabilization_rate_sp.q = 0;
+  }
+
+  if (YAW_RATE_DEADBAND_EXCEEDED() && !THROTTLE_STICK_DOWN()) {
+    stabilization_rate_sp.r = (float) radio_control.values[RADIO_YAW] * STABILIZATION_INDI_MAX_RATE / MAX_PPRZ;
+  } else {
+    stabilization_rate_sp.r = 0;
+  }
 }
 
 //Read rc with roll and yaw sitcks switched if the default orientation is vertical but airplane sticks are desired
 void stabilization_rate_read_rc_switched_sticks(void)
 {
-  //FIXME: make a new indi function
+  if (ROLL_RATE_DEADBAND_EXCEEDED()) {
+    stabilization_rate_sp.r =  - (float) radio_control.values[RADIO_ROLL] * STABILIZATION_INDI_MAX_RATE / MAX_PPRZ;
+  } else {
+    stabilization_rate_sp.r = 0;
+  }
+
+  if (PITCH_RATE_DEADBAND_EXCEEDED()) {
+    stabilization_rate_sp.q = (float) radio_control.values[RADIO_PITCH] * STABILIZATION_INDI_MAX_RATE / MAX_PPRZ;
+  } else {
+    stabilization_rate_sp.q = 0;
+  }
+
+  if (YAW_RATE_DEADBAND_EXCEEDED() && !THROTTLE_STICK_DOWN()) {
+    stabilization_rate_sp.p = (float) radio_control.values[RADIO_YAW] * STABILIZATION_INDI_MAX_RATE / MAX_PPRZ;
+  } else {
+    stabilization_rate_sp.p = 0;
+  }
+}
+
+void stabilization_rate_set_rpy_setpoint_i(struct Int32Eulers *rpy)
```

Comment:
Why do you need this function?

Reply:
To set the commanded rates coming from a different module. If you check the "stabilization_attitude_*.c" files, there is also a similar function.

Reply:
Ok, but then it is best to rename it in my opinion, because you're setting rates and not roll, pitch, yaw. Also, you are passing an Int32Eulers struct pointer but assigning to a FloatRates struct.

Reply:
Yes, you are right. I found in state.h file the definition for rates: "Int32Rates". I will change that and make sure that there is no type mismatch.

Reply:
Take care that Int32Rates != FloatRates

Reply:
Yes, I made the conversion using: RATES_FLOAT_OF_BFP ;)